### ISSUE TITLE
Change in behaviour for dont-query between pdns-recursor versions 4.3 and 4.4

### ISSUE THREAD

**Issue by arnoxit** [PowerDNS/pdns#10638]
<!-- Hi! Thanks for filing an issue. It will be read with care by human beings. Can we ask you to please fill out this template and not simply demand new features or send in complaints? Thanks! -->
<!-- Also please search the existing issues (both open and closed) to see if your report might be duplicate -->
<!-- Please don't file an issue when you have a support question, send support questions to the mailinglist or ask them on IRC (https://www.powerdns.com/opensource.html) -->
<!-- Before filing your ticket, please read our 'out in the open' support policy at https://blog.powerdns.com/2016/01/18/open-source-support-out-in-the-open/ -->

<!-- Tell us what is issue is about -->
 - Program: Recursor
 - Issue type: Bug report

### Short description
Forwarded requests to RFC1918 servers that are configured in forward-zone files are "`blocked by 'dont-query'`" setting. This is a change in behaviour compared to 4.3.7-1pdns.buster and to the documentation for the `dont-query` option which says:
_"Queries to addresses for zones as configured in any of the settings forward-zones, forward-zones-file or forward-zones-recurse are performed regardless of these limitations."_

### Environment
<!-- Tell us about the environment -->
 - Operating system: VyOS 1.4 (based on Debian Bullseye)
 - Software version: 4.4.2-3
 - Software source: VyOS repositories, sourced from Debian Bullseye repositories

### Steps to reproduce
<!-- Tell us step-by-step how the issue can be triggered. Please include your configuration files and any (Lua) scripts that are loaded. -->
1. Configure forward-zones-file with dns servers in a RFC-1918 address space (in our case 10.x.x.x/8)
2. Try and resolve an address within one of the forwarded domains

recursor.conf
```
### Autogenerated by dns_forwarding.py ###

# XXX: pdns recursor doesn't like whitespace near entry separators,
# especially in the semicolon-separated lists of name servers.
# Please be careful if you edit the template.

# Non-configurable defaults
daemon=yes
threads=1
allow-from=127.0.0.0/24
log-common-errors=yes
non-local-bind=yes
query-local-address=0.0.0.0
query-local-address6=::
lua-config-file=recursor.conf.lua

loglevel=9
trace=yes

# cache-size
max-cache-entries=10000

# negative TTL for NXDOMAIN
max-negative-ttl=5

# ignore-hosts-file
export-etc-hosts=yes

# listen-address
local-address=127.0.0.1

# dnssec
dnssec=process-no-validate

# serve rfc1918 records
serve-rfc1918=yes

forward-zones-file=recursor.forward-zones.conf
```

recursor.forward-zones.conf
```
# Autogenerated by VyOS (vyos-hostsd)
# Do not edit, your changes will get overwritten

# dot zone (catch-all): '+' indicates recursion is desired
# (same as forward-zones-recurse)
# static: 8.8.8.8, 1.1.1.1, 8.8.4.4, 1.0.0.1

+.=8.8.8.8, 1.1.1.1, 8.8.4.4, 1.0.0.1

# zones added via 'service dns forwarding domain'
0.10.in-addr.arpa=10.0.4.4, 10.0.104.4
100.10.in-addr.arpa=10.0.4.4, 10.0.104.4
example.net=10.0.4.4, 10.0.104.4
```

recursor.conf.lua
```
-- Autogenerated by VyOS (dns_forwarding.py) --
-- Do not edit, your changes will get overwritten --

-- Load DNSSEC root keys from dns-root-data package.
dofile("/usr/share/pdns-recursor/lua-config/rootkeys.lua")

-- Load lua from vyos-hostsd --
dofile("recursor.vyos-hostsd.conf.lua")
```

recursor.vyos-hostsd.conf.lua
```
-- Autogenerated by VyOS (vyos-hostsd) --
-- Do not edit, your changes will get overwritten --


-- from 'service dns forwarding domain'
```

### Expected behaviour
Address in the forwarded domain is resolved correctly as per previous version.

### Actual behaviour
Address is not resolved and trace log shows the following (client's domain name obfuscated):
```
$ dig node-1.test.example.net

; <<>> DiG 9.16.15-Debian <<>> node-1.test.example.net
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 45823
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;node-1.test.example.net. IN A

;; Query time: 61 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Mon Aug 09 19:51:40 UTC 2021
;; MSG SIZE  rcvd: 66
```
trace
```
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=(empty): doResolve
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Wants DNSSEC processing, auth data in query for A 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Recursion not requested for 'node-1.test.example.net|A', peeking at auth/forward zones
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Looking for CNAME cache hit of 'node-1.test.example.net|CNAME'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Looking for DNAME cache hit of 'node-1.test.example.net|DNAME' or its ancestors
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: No CNAME or DNAME cache hit of 'node-1.test.example.net' found
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=(empty): Step0 Not cached
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=(empty): Step0 qname is in a forwarded domain example.net 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Checking if we have NS in cache for 'node-1.test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: no valid/useful NS in cache for 'node-1.test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Checking if we have NS in cache for 'test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: no valid/useful NS in cache for 'test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=test.example.net: Step2 New child
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=test.example.net: Step4 Resolve A for child 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Wants DNSSEC processing, auth data in query for A
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Looking for CNAME cache hit of 'test.example.net|CNAME'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Looking for DNAME cache hit of 'test.example.net|DNAME' or its ancestors
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: No CNAME or DNAME cache hit of 'test.example.net' found 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: No cache hit for 'test.example.net|A', trying to find an appropriate NS record
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : got TA for '.'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : setting cut state for . to Secure
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Checking if we have NS in cache for 'test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: no valid/useful NS in cache for 'test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Checking if we have NS in cache for 'example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: no valid/useful NS in cache for 'example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Checking if we have NS in cache for 'net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (e.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (d.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (i.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (g.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (c.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (h.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (b.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (j.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (f.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (a.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (l.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (m.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS in cache for 'net', but needs glue (k.gtld-servers.net) which we miss or is expired
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: no valid/useful NS in cache for 'net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Checking if we have NS in cache for '.' 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'a.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'b.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'c.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'd.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'e.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'f.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'g.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'h.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'i.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'j.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'k.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'l.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: NS (with ip, or non-glue) in cache for '.' -> 'm.root-servers.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: within bailiwick: 1,  in cache, ttl=3599965
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: We have NS in cache for '.' (flawedNSSet=0)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net authOrForwDomain: example.net nsFromCacheDomain: . isPartof: 1
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: using forwarder as NS
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: initial validation status for test.example.net is Indeterminate
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Cache consultations done, have 1 NS to contact
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Domain has hardcoded nameservers
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net.: Nameservers: -10.0.4.4:53(47.13ms), -10.0.104.4:53(109.39ms)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Resolved 'example.net' NS (empty) to: 10.0.4.4, 10.0.104.4
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Trying IP 10.0.4.4:53, asking 'test.example.net|A'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: Got 6 answers from (empty) (10.0.4.4), rcode=0 (No Error), aa=1, in 58ms
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: accept answer 'test.example.net|A|10.0.1.83' from 'example.net' nameservers? ttl=86400, place=1 YES! - This answer was received from a server we forward to.
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: accept answer 'test.example.net|NS|idm-2.internal.example.net.' from 'example.net' nameservers? ttl=86400, place=2 YES! - This answer was received from a server we forward to.
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: accept answer 'test.example.net|NS|idm-1.internal.example.net.' from 'example.net' nameservers? ttl=86400, place=2 YES! - This answer was received from a server we forward to.
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: accept answer 'idm-1.internal.example.net|A|10.0.4.4' from 'example.net' nameservers? ttl=1200, place=3 YES! - This answer was received from a server we forward to.
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: accept answer 'idm-2.internal.example.net|A|10.0.104.4' from 'example.net' nameservers? ttl=1200, place=3 YES! - This answer was received from a server we forward to.
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: OPT answer '.' from 'example.net' nameservers
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : got initial zone status Indeterminate for record test.example.net|A
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : got initial zone status Indeterminate for record idm-2.internal.example.net|A
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : got initial zone status Indeterminate for record idm-1.internal.example.net|A
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : got initial zone status Indeterminate for record test.example.net|NS
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: determining status after receiving this packet
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: answer is in: resolved to '10.0.1.83|A'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: got NS record 'test.example.net' -> 'idm-2.internal.example.net.'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: got NS record 'test.example.net' -> 'idm-1.internal.example.net.'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: status=got results, this level of recursion done
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] test.example.net: validation status is Indeterminate
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=test.example.net: Step4 Resolve A result is No Error/1/1
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=node-1.test.example.net: Step2 New child
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=node-1.test.example.net: Step3 Going to do final resolve
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Wants DNSSEC processing, auth data in query for A 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Looking for CNAME cache hit of 'node-1.test.example.net|CNAME'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Looking for DNAME cache hit of 'node-1.test.example.net|DNAME' or its ancestors
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: No CNAME or DNAME cache hit of 'node-1.test.example.net' found
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: No cache hit for 'node-1.test.example.net|A', trying to find an appropriate NS record
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : got TA for '.'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] : setting cut state for . to Secure
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Checking if we have NS in cache for 'node-1.test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: no valid/useful NS in cache for 'node-1.test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Checking if we have NS in cache for 'test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: NS (with ip, or non-glue) in cache for 'test.example.net' -> 'idm-2.internal.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: within bailiwick: 0, not in cache / did not look at cache
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: NS (with ip, or non-glue) in cache for 'test.example.net' -> 'idm-1.internal.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: within bailiwick: 0, not in cache / did not look at cache
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: We have NS in cache for 'test.example.net' (flawedNSSet=0)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net authOrForwDomain: example.net nsFromCacheDomain: test.example.net isPartof: 0 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: using NS from cache
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: initial validation status for node-1.test.example.net is Indeterminate
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Cache consultations done, have 2 NS to contact
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net.: Nameservers: idm-1.internal.example.net(0.00ms), idm-2.internal.example.net(0.00ms)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Trying to resolve NS 'idm-1.internal.example.net' (1/2)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  QM idm-1.internal.example.net.|A child=(empty): doResolve
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: Wants DNSSEC processing, NO auth data in query for A
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: Recursion not requested for 'idm-1.internal.example.net|A', peeking at auth/forward zones
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: Looking for CNAME cache hit of 'idm-1.internal.example.net|CNAME'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: Looking for DNAME cache hit of 'idm-1.internal.example.net|DNAME' or its ancestors
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: No CNAME or DNAME cache hit of 'idm-1.internal.example.net' found
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: Found cache hit for A: 10.0.4.4[ttl=1200]
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-1.internal.example.net: updating validation state with cache content for idm-1.internal.example.net to Indeterminate
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  QM idm-1.internal.example.net.|A child=(empty): Step0 Found in cache
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Resolved 'test.example.net' NS idm-1.internal.example.net to: 10.0.4.4
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Trying IP 10.0.4.4:53, asking 'node-1.test.example.net|A'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: not sending query to 10.0.4.4, blocked by 'dont-query' setting
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Trying to resolve NS 'idm-2.internal.example.net' (2/2)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  QM idm-2.internal.example.net.|A child=(empty): doResolve
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: Wants DNSSEC processing, NO auth data in query for A
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: Recursion not requested for 'idm-2.internal.example.net|A', peeking at auth/forward zones 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: Looking for CNAME cache hit of 'idm-2.internal.example.net|CNAME'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: Looking for DNAME cache hit of 'idm-2.internal.example.net|DNAME' or its ancestors
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: No CNAME or DNAME cache hit of 'idm-2.internal.example.net' found 
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: Found cache hit for A: 10.0.104.4[ttl=1200]
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  idm-2.internal.example.net: updating validation state with cache content for idm-2.internal.example.net to Indeterminate
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3]  QM idm-2.internal.example.net.|A child=(empty): Step0 Found in cache
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Resolved 'test.example.net' NS idm-2.internal.example.net to: 10.0.104.4
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Trying IP 10.0.104.4:53, asking 'node-1.test.example.net|A'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: not sending query to 10.0.104.4, blocked by 'dont-query' setting
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: Failed to resolve via any of the 2 offered NS at level 'test.example.net'
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] node-1.test.example.net: failed (res=-1)
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: [3] QM node-1.test.example.net.|A child=node-1.test.example.net: Step3 Final resolve: Server Failure/0
Aug 09 19:51:40 vyos-1.remote.example.net pdns_recursor[3463]: 1 [3/1] answer to question 'node-1.test.example.net|A': 0 answers, 1 additional, took 1 packets, 58.095 netw ms, 59.696 tot ms, 0 throttled, 0 timeouts, 0 tcp connections, rcode=2

```
### Other information
<!-- if you already did more digging into the issue, please provide all the information you gathered -->
On a Vyos 1.4 host based on Debian Buster with pdns-recursor 4.3.7-1pdns-buster and the same configuration the behaviour is different and matches expectation and the documentation:

```
vyos@vyos-2.remote.example.net:~$ dig node-1.test.example.net                
                                                                                                                                                                                                                                                                                                                                                                                             
; <<>> DiG 9.11.5-P4-5.1+deb10u5-Debian <<>> node-1.test.example.net                                                                                                                                                                                                                                                                                                                         
;; global options: +cmd                                                                                                                                                                                                                                                                                                                                                                      
;; Got answer:                                                                                                                                                                                                                                                                                                                                                                               
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 27933                                                                                                                                                                                                                                                                                                                                    
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1                                                                                                                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                                                                                             
;; OPT PSEUDOSECTION:                                                                                                                                                                                                                                                                                                                                                                        
; EDNS: version: 0, flags:; udp: 512                                                                                                                                                                                                                                                                                                                                                         
;; QUESTION SECTION:                                                                                                                                                                                                                                                                                                                                                                         
;node-1.test.example.net. IN A                                                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                                                                                             
;; ANSWER SECTION:                                                                                                                                                                                                                                                                                                                                                                           
node-1.test.example.net. 1200 IN A 10.0.1.183                                                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                                                                                             
;; Query time: 51 msec                                                                                                                                                                                                                                                                                                                                                                       
;; SERVER: 127.0.0.1#53(127.0.0.1)                                                                                                                                                                                                                                                                                                                                                           
;; WHEN: Mon Aug 09 20:02:00 UTC 2021                                                                                                                                                                                                                                                                                                                                                        
;; MSG SIZE  rcvd: 82                                                                
```

trace
```
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: 1 [452/1] question for 'node-1.test.example.net|A' from 127.0.0.1:46074                                                                                                                                                                                                                    
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Wants DNSSEC processing, auth data in query for A                                                                                                                                                                                                                                             
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Looking for CNAME cache hit of 'node-1.test.example.net|CNAME'                                                                                                                                                                                                                                
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Looking for DNAME cache hit of 'node-1.test.example.net|DNAME' or its ancestors                                                                                                                                                                                                               
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: No CNAME or DNAME cache hit of 'node-1.test.example.net' found                                                                                                                                                                                                                                
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: No cache hit for 'node-1.test.example.net|A', trying to find an appropriate NS record                                                                                                                                                                                                         
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] : got TA for '.'                                                                                                        
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] : setting cut state for . to Secure                                                                                                                                                                                                                                                                                    
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: initial validation status for node-1.test.example.net is Indeterminate
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Cache consultations done, have 1 NS to contact
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Domain has hardcoded nameservers                                                               
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net.: Nameservers: -10.0.104.4:53(0.00ms), -10.0.4.4:53(46.84ms)                                                                                                                                                                                                                                   
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Resolved 'example.net' NS (empty) to: 10.0.104.4, 10.0.4.4                                                                                                                                                                                                                                    
Aug 09 20:01:59 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Trying IP 10.0.104.4:53, asking 'node-1.test.example.net|A'                                                                                                                                                                                                                                   
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: Got 6 answers from (empty) (10.0.104.4), rcode=0 (No Error), aa=1, in 48ms                                                                                                                                                                                                                    
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: accept answer 'node-1.test.example.net|A|10.0.1.183' from 'example.net' nameservers? ttl=1200, place=1 YES! - This answer was received from a server we forward to.                                                                                                                           
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: accept answer 'test.example.net|NS|idm-2.internal.example.net.' from 'example.net' nameservers? ttl=86400, place=2 YES! - This answer was received from a server we forward to.                                                                                                               
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: accept answer 'test.example.net|NS|idm-1.internal.example.net.' from 'example.net' nameservers? ttl=86400, place=2 YES! - This answer was received from a server we forward to.
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: accept answer 'idm-1.internal.example.net|A|10.0.4.4' from 'example.net' nameservers? ttl=1200, place=3 YES! - This answer was received from a server we forward to.
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: accept answer 'idm-2.internal.example.net|A|10.0.104.4' from 'example.net' nameservers? ttl=1200, place=3 YES! - This answer was received from a server we forward to.
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: OPT answer '.' from 'example.net' nameservers                                                  
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] : got initial zone status Indeterminate for record node-1.test.example.net|A                                                                                                                                                                                                                                           
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] : got initial zone status Indeterminate for record idm-2.internal.example.net|A                                         
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] : got initial zone status Indeterminate for record idm-1.internal.example.net|A                                                                                                                                                                                                                                        
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] : got initial zone status Indeterminate for record test.example.net|NS                                                  
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: determining status after receiving this packet                                                 
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: answer is in: resolved to '10.0.1.183|A'                                                                                                                                                                                                                                                      
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: got NS record 'test.example.net' -> 'idm-2.internal.example.net.'
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: got NS record 'test.example.net' -> 'idm-1.internal.example.net.' 
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: status=got results, this level of recursion done                                                                                                                                                                                                                                              
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: [452] node-1.test.example.net: validation status is Indeterminate                                                             
Aug 09 20:02:00 vyos-2.remote.example.net pdns_recursor[10432]: 1 [452/1] answer to question 'node-1.test.example.net|A': 1 answers, 1 additional, took 1 packets, 48.952 netw ms, 49.324 tot ms, 0 throttled, 0 timeouts, 0 tcp connections, rcode=0                                                                     
```

**Comment by omoerbeek** [comment]
Looks like a duplicate of #10533. Will test later today.

**Comment by omoerbeek** [comment]
OK the analysis so far.  This is not exactly the same thing as #10533. 

1. Likely qname minimizationm does not matter in this case. Please confirm.
 
2. What happens is, that since 4.4, cached NS records for a subdomain inside a forwarded domain get priority above the forward statement. This is needed to allow e.g. a local root zone.
But if the cached entry is used, the property "this is a forwarded domain" is not set, so the code that contacts the nameserver does not regards this as a special forwarded case and the `dont-query` rule kicks in.

Sadly the `this is a nameserver IP for a forwarded domain` is an implicit property (absence of name) and that makes it a bit harder to take that into account.

Some likely workarounds:
1. Get rid of the don't query line in the config. This might not be possible for your use-case.
2. List the subdomain also as a forwarded domain explicitly. e.g.
```
example.net=10.0.4.4, 10.0.104.4
test.example.net=10.0.4.4, 10.0.104.4
```
forward statements that match a domain exactly get precedence compared to cache contents.

I have to think how to solve this in a general way.

**Comment by arnoxit** [comment]
Thanks for looking into it.

I tried `qname-minimization=no` but saw the same issue. It is interesting that it seems to manage to resolve the first entry after restarting with the new option and then fails thereafter... which I think confirms your point 2, with the first query not coming from the cache but further queries hitting the issue.

We will see if we can work out a way of excluding our range from the `dont-query` within the mechanism of the VyOS config, or we may ask them to pin the version of pdns-recursor back to the earlier version if not. We would not want to add all the subdomains to the config as there are many and it would require updating all the firewalls every time a subdomain is added or removed.

**Comment by omoerbeek** [comment]
Please don't pin to old versions if at all possible. 4.3.x will be EOL this fall. I hope to come up with a proper fix the coming days.
